modules/ingester/instance_search.go

```diff
@@ -149,20 +75,93 @@ func (i *instance) searchBlock(ctx context.Context, req *tempopb.SearchRequest,
 		}
 		if err != nil {
 			level.Error(log.Logger).Log("msg", "error searching block", "blockID", blockID, "err", err)
-			sr.SetError(err)
+			anyErr.Store(err)
+			return
+		}
+
+		if resp == nil {
 			return
 		}
 
-		for _, t := range resp.Traces {
-			if sr.AddResult(ctx, t) {
-				break
+		resultsMtx.Lock()
+		defer resultsMtx.Unlock()
+
+		if resp.Metrics != nil {
+			metrics.InspectedTraces += resp.Metrics.InspectedTraces
+			metrics.InspectedBytes += resp.Metrics.InspectedBytes
+		}
+
+		for _, tr := range resp.Traces {
+			combiner.AddMetadata(tr)
+			if combiner.Count() >= maxResults {
+				return
 			}
 		}
-		sr.AddBlockInspected()
+	}
 
-		sr.AddBytesInspected(resp.Metrics.InspectedBytes)
-		sr.AddTraceInspected(resp.Metrics.InspectedTraces)
-	}(block, cleanup)
+	// Search headblock (synchronously)
+	// Lock headblock separately from other blocks and release it as quickly as possible.
```

Comment:
i only see one mutex locked (unless this comment is also referring to the resultsMtx). is this still relevant?

Reply:
The rest of the blocks are locked [here](https://github.com/grafana/tempo/pull/3328/files#diff-8c9aac72ec16ad39eb093753942cde7f893be0b5950578ffe10ef003d2ff5237R128)

There are 3 mutexes:
* i.headBlockMtx - for head block, unlocked asap
* i.blocksMtx - for all completing and complete blocks, unlocked when the last goroutine exits
* resultsMtx - for the concurrent results

Reply:
since they aren't nested, does the order matter? wouldn't the headblock be released and then it would sit waiting for the blocks mutex? if another goroutine did it in the opposite order they would unblock for each other?

ultimately it's not a big deal. i just read this comment and it seemed that the order was unimportant.